1. Write a Pandas program to find and replace the missing values in a given DataFrame which do not have any valuable information. <br>
Hint: Use learnt np.nan values where it is NaN or similar in the given table.

Example: <br>
Missing values: ?, -- <br>
Replace those values with NaN <br>

Test Data: <br>

 ord_no, purch_amt, ord_date, customer_id, salesman_id <br>
0, 70001, 150.5, ?, 3002, 5002 <br>
1, NaN, 270.65, 2012-09-10, 3001, 5003 <br>
2, 70002, 65.26, NaN, 3001, ? <br>
3, 70004, 110.5, 2012-08-17, 3003, 5001 <br>
4, NaN, 948.5, 2012-09-10, 3002, NaN <br>
5, 70005, 2400.6, 2012-07-27, 3001, 5002 <br>
6, --, 5760, 2012-09-10, 3001, 5001 <br>
7, 70010, ?, 2012-10-10, 3004, ? <br>
8, 70003, 12.43, 2012-10-10, --, 5003 <br>
9, 70012, 2480.4, 2012-06-27, 3002, 5002 <br>
10, NaN, 250.45, 2012-08-17, 3001, 5003 <br>
11, 70013, 3045.6, 2012-04-25, 3001, --

In [2]:
import pandas as pd
import numpy as np
df = pd.DataFrame({
    'ord_no': [70001, np.nan, 70002, 70004, np.nan, 70005, '--', 70010, 70003, 70012, np.nan, 70013],
    'purch_amt': [150.5, 270.65, 65.26, 110.5, 948.5, 2400.6, 5760, '?', 12.43, 2480.4, 250.45, 3045.6],
    'ord_date': ['?','2012-09-10',np.nan,'2012-08-17','2012-09-10','2012-07-27','2012-09-10','2012-10-10','2012-10-10','2012-06-27','2012-08-17','2012-04-25'],
    'custormer_id': [3002, 3001, 3001, 3003, 3002, 3001, 3001, 3004, '--', 3002, 3001, 3001],
    'salesman_id': [5002, 5003, '?', 5001, np.nan, 5002, 5001, '?', 5003, 5002, 5003, '--']})
result = df.replace({'?': np.nan, '--': np.nan})
result
    

ord_no  purch_amt    ord_date  custormer_id  salesman_id
0   70001.0     150.50         NaN        3002.0       5002.0
1       NaN     270.65  2012-09-10        3001.0       5003.0
2   70002.0      65.26         NaN        3001.0          NaN
3   70004.0     110.50  2012-08-17        3003.0       5001.0
4       NaN     948.50  2012-09-10        3002.0          NaN
5   70005.0    2400.60  2012-07-27        3001.0       5002.0
6       NaN    5760.00  2012-09-10        3001.0       5001.0
7   70010.0        NaN  2012-10-10        3004.0          NaN
8   70003.0      12.43  2012-10-10           NaN       5003.0
9   70012.0    2480.40  2012-06-27        3002.0       5002.0
10      NaN     250.45  2012-08-17        3001.0       5003.0
11  70013.0    3045.60  2012-04-25        3001.0          NaN

2. Write a Pandas program to drop those rows from a given DataFrame in which specific columns have missing values. Hint: You'll need 'subset' keyword

Test data:

ord_no, purch_amt, ord_date, customer_id <br>
0, NaN, NaN, NaN, NaN <br>
1, NaN, 270.65, 2012-09-10, 3001.0 <br>
2, 70002.0, 65.26, NaN, 3001.0 <br>
3, NaN, NaN, NaN, NaN <br>
4, NaN, 948.50, 2012-09-10, 3002.0 <br>
5, 70005.0, 2400.60, 2012-07-27, 3001.0 <br>
6, NaN, 5760.00, 2012-09-10, 3001.0 <br>
7, 70010.0, 1983.43, 2012-10-10, 3004.0 <br>
8, 70003.0, 2480.40,  2012-10-10, 3003.0 <br>
9, 70012.0, 250.45, 2012-06-27, 3002.0 <br
10, NaN, 75.29, 2012-08-17,  3001.0 <br>
11, NaN, NaN, NaN, NaN

In [4]:
import pandas as pd
import numpy as np
df = pd.DataFrame({
    'ord_no':[np.nan,np.nan,70002,np.nan,np.nan,70005,np.nan,70010,70003,70012,np.nan,np.nan],
    'purch_amt':[np.nan,270.65,65.26,np.nan,948.5,2400.6,5760,1983.43,2480.4,250.45, 75.29,np.nan],
    'ord_date': [np.nan,'2012-09-10',np.nan,np.nan,'2012-09-10','2012-07-27','2012-09-10','2012-10-10','2012-10-10','2012-06-27','2012-08-17',np.nan],
    'customer_id':[np.nan,3001,3001,np.nan,3002,3001,3001,3004,3003,3002,3001,np.nan]})
result = df.dropna(subset=['ord_no', 'ord_date', 'customer_id'])
result

ord_no  purch_amt    ord_date  customer_id
5  70005.0    2400.60  2012-07-27       3001.0
7  70010.0    1983.43  2012-10-10       3004.0
8  70003.0    2480.40  2012-10-10       3003.0
9  70012.0     250.45  2012-06-27       3002.0

3. You developed a multiplayer indie game called 🌹 Rose Thorn. Players compete in one of two venues - the ocean or the desert. You track the outcome of five games between three players in a DataFrame called games.

In [6]:
import numpy as np
import pandas as pd

games = pd.DataFrame({
    'bella1':   ['2nd', '3rd', '1st', '2nd', '3rd'],
    'billybob': ['1st', '2nd', '2nd', '1st', '2nd'],
    'nosoup4u': ['3rd', '1st', '3rd', '3rd', '3rd'],
    'venue': ['desert', 'ocean', 'desert', 'ocean', 'desert']
})

games

bella1 billybob nosoup4u   venue
0    2nd      1st      3rd  desert
1    3rd      2nd      1st   ocean
2    1st      2nd      3rd  desert
3    2nd      1st      3rd   ocean
4    3rd      2nd      3rd  desert

Now you want to analyze the data. Convert the `games` DataFrame into a new DataFrame that identifies how many times each (`player`, `placement`) occurs per `venue`, specifically with `venue` as the row index and (`player`, `placed`) as the column MultiIndex.

In [7]:
# 1. The first step is to use the DataFrame.melt() function to "stack" the columns bella1, billybob, and nosoup4u into a single column.

step1 = games.melt(id_vars='venue', var_name='player', value_name='placed')
step1

venue    player placed
0   desert    bella1    2nd
1    ocean    bella1    3rd
2   desert    bella1    1st
3    ocean    bella1    2nd
4   desert    bella1    3rd
5   desert  billybob    1st
6    ocean  billybob    2nd
7   desert  billybob    2nd
8    ocean  billybob    1st
9   desert  billybob    2nd
10  desert  nosoup4u    3rd
11   ocean  nosoup4u    1st
12  desert  nosoup4u    3rd
13   ocean  nosoup4u    3rd
14  desert  nosoup4u    3rd

In [8]:
# 2. Next, if we group the data by (venue, player, placed), then aggregate the groups with 'size', we get back the count of rows in each group. This tells us how many times each player got each placement for each venue, amongst the triplets that actually occurred in the data.

step2 = step1.groupby(['venue', 'player', 'placed']).agg('size')
step2

venue   player    placed
desert  bella1    1st       1
                  2nd       1
                  3rd       1
        billybob  1st       1
                  2nd       2
        nosoup4u  3rd       3
ocean   bella1    2nd       1
                  3rd       1
        billybob  1st       1
                  2nd       1
        nosoup4u  1st       1
                  3rd       1
dtype: int64

In [9]:
# 3. However, we also want triplets that didn't occur, like 
#(desert, billybob, 3rd) and (ocean, nosoup4u, 2nd). 
#There's a few ways can fix this, but one really cool way is, 
#let's got back to step1 convert 'venue', 'player', and 'placed' to Categoricals.

step1['venue'] = pd.Categorical(step1.venue)
step1['player'] = pd.Categorical(step1.player)
step1['placed'] = pd.Categorical(step1.placed)

In [10]:
# 4. Now when we group the data in step2, all those unobserved triplets are included in the result!

step2 = step1.groupby(['venue', 'player', 'placed']).agg('size')
step2

venue   player    placed
desert  bella1    1st       1
                  2nd       1
                  3rd       1
        billybob  1st       1
                  2nd       2
                  3rd       0
        nosoup4u  1st       0
                  2nd       0
                  3rd       3
ocean   bella1    1st       0
                  2nd       1
                  3rd       1
        billybob  1st       1
                  2nd       1
                  3rd       0
        nosoup4u  1st       1
                  2nd       0
                  3rd       1
dtype: int64

In [11]:
# 5. At this point, step2 is a dataframe with a 3-level multiindex. 
#If we reshape it using unstack(), with level=['player','placed'] 
#to move the player and placed levels out of the row index and 
#into the column index, we get back our desired result.

step3 = step2.unstack(level=['player', 'placed'])
step3

player bella1         billybob         nosoup4u        
placed    1st 2nd 3rd      1st 2nd 3rd      1st 2nd 3rd
venue                                                  
desert      1   1   1        1   2   0        0   0   3
ocean       0   1   1        1   1   0        1   0   1

4. You run an ecommerce site called shoesfordogs.com . You want to analyze your visitors, so you compile a DataFrame called hits that represents each time a visitor hit some page on your site.

In [15]:
import numpy as np
import pandas as pd

# exposed places
exposed = [0,5,9]

# whereabouts of each person
generator = np.random.default_rng(2468)
Nplaces = 10
Npersons = 10
place_ids = np.arange(Nplaces)
visits = generator.choice(place_ids, size=3*Nplaces, replace=True)
split_idxs = np.sort(generator.choice(len(visits), size=9, replace=True))
whereabouts = pd.DataFrame({
    'person_id': range(Npersons),
    'places': [np.unique(x).tolist() for x in np.array_split(visits, split_idxs)]
})

whereabouts

person_id              places
0          0        [3, 4, 5, 6]
1          1                  []
2          2                 [3]
3          3           [6, 8, 9]
4          4                 [3]
5          5  [0, 2, 5, 6, 7, 8]
6          6              [2, 7]
7          7        [0, 5, 8, 9]
8          8           [2, 7, 9]
9          9        [0, 5, 8, 9]

For each `person`, identify the `places` they visited which have been `exposed`. Make this a new list-column in `whereabouts` called `exposures`.


In [14]:
#SOLUTION 1
whereabouts['exposures'] = whereabouts['places'].apply(lambda x: list(set(x) & set(exposed)))
whereabouts

person_id              places  exposures
0          0        [3, 4, 5, 6]        [5]
1          1                  []         []
2          2                 [3]         []
3          3           [6, 8, 9]        [9]
4          4                 [3]         []
5          5  [0, 2, 5, 6, 7, 8]     [0, 5]
6          6              [2, 7]         []
7          7        [0, 5, 8, 9]  [0, 9, 5]
8          8           [2, 7, 9]        [9]
9          9        [0, 5, 8, 9]  [0, 9, 5]

In [16]:
# SOLUTION 2
# 1. Expand whereabouts. We use the DataFrame.explode() method with column='places'.
expanded = whereabouts.explode(column='places')
expanded

person_id places
0          0      3
0          0      4
0          0      5
0          0      6
1          1    NaN
2          2      3
3          3      6
3          3      8
3          3      9
4          4      3
5          5      0
5          5      2
5          5      5
5          5      6
5          5      7
5          5      8
6          6      2
6          6      7
7          7      0
7          7      5
7          7      8
7          7      9
8          8      2
8          8      7
8          8      9
9          9      0
9          9      5
9          9      8
9          9      9

In [18]:
# 2. Select rows where places is one of the exposed places.
filtered = expanded.loc[expanded.places.isin(exposed)]
filtered

person_id places
0          0      5
3          3      9
5          5      0
5          5      5
7          7      0
7          7      5
7          7      9
8          8      9
9          9      0
9          9      5
9          9      9

In [19]:
# 3. Group the data by person_id, aggregating the exposed places with a list.
aggregated = filtered.groupby('person_id')[['places']].agg(list)
aggregated

places
person_id           
0                [5]
3                [9]
5             [0, 5]
7          [0, 5, 9]
8                [9]
9          [0, 5, 9]

In [21]:
# 4. Rename the column 'places' to 'exposures'.
aggregated.rename(columns={'places': 'exposures'}, inplace=True)
aggregated

exposures
person_id           
0                [5]
3                [9]
5             [0, 5]
7          [0, 5, 9]
8                [9]
9          [0, 5, 9]

In [22]:
# 5. Insert this column of exposures into the whereabouts DataFrame. A left join does the trick.
whereabouts = pd.merge(left=whereabouts, right=aggregated, how='left', on='person_id')
whereabouts

person_id              places  exposures
0          0        [3, 4, 5, 6]        [5]
1          1                  []        NaN
2          2                 [3]        NaN
3          3           [6, 8, 9]        [9]
4          4                 [3]        NaN
5          5  [0, 2, 5, 6, 7, 8]     [0, 5]
6          6              [2, 7]        NaN
7          7        [0, 5, 8, 9]  [0, 5, 9]
8          8           [2, 7, 9]        [9]
9          9        [0, 5, 8, 9]  [0, 5, 9]

In [23]:
# 6. Replace NaNs with empty lists.
# Unfortunately you can't use fillna() to fill NaNs with an empty list. So, we have to do something like this..
# Here, lambda x: returns x if it's already a list, otherwise it returns an empty list [].
whereabouts['exposures'] = whereabouts.exposures.apply(lambda x: x if isinstance(x, list) else [])
whereabouts

person_id              places  exposures
0          0        [3, 4, 5, 6]        [5]
1          1                  []         []
2          2                 [3]         []
3          3           [6, 8, 9]        [9]
4          4                 [3]         []
5          5  [0, 2, 5, 6, 7, 8]     [0, 5]
6          6              [2, 7]         []
7          7        [0, 5, 8, 9]  [0, 5, 9]
8          8           [2, 7, 9]        [9]
9          9        [0, 5, 8, 9]  [0, 5, 9]

5. You run an ecommerce site called shoesfordogs.com . You want to analyze your visitors, so you compile a DataFrame called hits that represents each time a visitor hit some page on your site. 

In [1]:
import numpy as np
import pandas as pd

pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

generator = np.random.default_rng(90)
products = ['iev','pys','vae','dah','yck','axl','apx','evu','wqv','tfg','aur','rgy','kef','lzj','kiz','oma']
hits = pd.DataFrame({
    'visitor_id': generator.choice(5, size=20, replace=True) + 1,
    'session_id': generator.choice(4, size=20, replace=True),
    'date_time': pd.to_datetime('2020-01-01') + pd.to_timedelta(generator.choice(60, size=20), unit='m'),
    'page_url': [f'shoesfordogs.com/product/{x}' for x in generator.choice(products, size=20, replace=True)]
})
hits['session_id'] = hits.visitor_id * 100 + hits.session_id

You suspect that the undocumented third-party tracking system on your website is buggy and sometimes splits one session into two or more session_ids. You want to correct this behavior by creating a field called session_group_id that stitches broken session_ids together.

Two session, A & B, should belong to the same session group if <br>
 - They have the same visitor_id and <br>
    - Their hits overlap in time or <br>
    - The latest hit from A is within five minutes of the earliest hit from B, or vice-versa <br>

<b>Associativity applies.</b> So, if A is grouped with B, and B is grouped with C, then A should be grouped with C as well.

Create a column in hits called session_group_id that identifies which hits belong to the same session group.

In [2]:
hits.sort_values(by=['visitor_id', 'date_time'], inplace=True)
hits

visitor_id  session_id           date_time                      page_url
19           1         100 2020-01-01 00:11:00  shoesfordogs.com/product/apx
2            1         102 2020-01-01 00:48:00  shoesfordogs.com/product/evu
11           1         101 2020-01-01 00:52:00  shoesfordogs.com/product/iev
18           2         200 2020-01-01 00:16:00  shoesfordogs.com/product/aur
1            2         200 2020-01-01 00:18:00  shoesfordogs.com/product/oma
6            2         203 2020-01-01 00:37:00  shoesfordogs.com/product/rgy
4            2         201 2020-01-01 00:40:00  shoesfordogs.com/product/yck
12           3         302 2020-01-01 00:01:00  shoesfordogs.com/product/dah
17           3         303 2020-01-01 00:01:00  shoesfordogs.com/product/lzj
5            3         302 2020-01-01 00:33:00  shoesfordogs.com/product/pys
8            3         302 2020-01-01 00:48:00  shoesfordogs.com/product/kef
10           3         300 2020-01-01 00:49:00  shoesfordogs.com/product/kef
7            3         302 2020-01-01 00:54:00  shoesfordogs.com/product/tfg
13           4         403 2020-01-01 00:02:00  shoesfordogs.com/product/lzj
0            4         400 2020-01-01 00:05:00  shoesfordogs.com/product/pys
3            4         403 2020-01-01 00:21:00  shoesfordogs.com/product/oma
9            4         402 2020-01-01 00:24:00  shoesfordogs.com/product/apx
14           4         401 2020-01-01 00:42:00  shoesfordogs.com/product/evu
16           5         503 2020-01-01 00:31:00  shoesfordogs.com/product/dah
15           5         500 2020-01-01 00:39:00  shoesfordogs.com/product/apx

In [3]:
sessions = hits.groupby(['visitor_id', 'session_id']).agg(
    date_time_min=('date_time', 'min'),
    date_time_max=('date_time', 'max')
)
sessions

date_time_min       date_time_max
visitor_id session_id                                        
1          100        2020-01-01 00:11:00 2020-01-01 00:11:00
           101        2020-01-01 00:52:00 2020-01-01 00:52:00
           102        2020-01-01 00:48:00 2020-01-01 00:48:00
2          200        2020-01-01 00:16:00 2020-01-01 00:18:00
           201        2020-01-01 00:40:00 2020-01-01 00:40:00
           203        2020-01-01 00:37:00 2020-01-01 00:37:00
3          300        2020-01-01 00:49:00 2020-01-01 00:49:00
           302        2020-01-01 00:01:00 2020-01-01 00:54:00
           303        2020-01-01 00:01:00 2020-01-01 00:01:00
4          400        2020-01-01 00:05:00 2020-01-01 00:05:00
           401        2020-01-01 00:42:00 2020-01-01 00:42:00
           402        2020-01-01 00:24:00 2020-01-01 00:24:00
           403        2020-01-01 00:02:00 2020-01-01 00:21:00
5          500        2020-01-01 00:39:00 2020-01-01 00:39:00
           503        2020-01-01 00:31:00 2020-01-01 00:31:00

In [5]:
sessions.sort_values(['visitor_id', 'date_time_min'], inplace=True)
sessions

date_time_min       date_time_max
visitor_id session_id                                        
1          100        2020-01-01 00:11:00 2020-01-01 00:11:00
           102        2020-01-01 00:48:00 2020-01-01 00:48:00
           101        2020-01-01 00:52:00 2020-01-01 00:52:00
2          200        2020-01-01 00:16:00 2020-01-01 00:18:00
           203        2020-01-01 00:37:00 2020-01-01 00:37:00
           201        2020-01-01 00:40:00 2020-01-01 00:40:00
3          302        2020-01-01 00:01:00 2020-01-01 00:54:00
           303        2020-01-01 00:01:00 2020-01-01 00:01:00
           300        2020-01-01 00:49:00 2020-01-01 00:49:00
4          403        2020-01-01 00:02:00 2020-01-01 00:21:00
           400        2020-01-01 00:05:00 2020-01-01 00:05:00
           402        2020-01-01 00:24:00 2020-01-01 00:24:00
           401        2020-01-01 00:42:00 2020-01-01 00:42:00
5          503        2020-01-01 00:31:00 2020-01-01 00:31:00
           500        2020-01-01 00:39:00 2020-01-01 00:39:00

In [6]:
sessions['date_time_max_p5'] = sessions.date_time_max + pd.Timedelta(minutes=5)
sessions

date_time_min       date_time_max    date_time_max_p5
visitor_id session_id                                                            
1          100        2020-01-01 00:11:00 2020-01-01 00:11:00 2020-01-01 00:16:00
           102        2020-01-01 00:48:00 2020-01-01 00:48:00 2020-01-01 00:53:00
           101        2020-01-01 00:52:00 2020-01-01 00:52:00 2020-01-01 00:57:00
2          200        2020-01-01 00:16:00 2020-01-01 00:18:00 2020-01-01 00:23:00
           203        2020-01-01 00:37:00 2020-01-01 00:37:00 2020-01-01 00:42:00
           201        2020-01-01 00:40:00 2020-01-01 00:40:00 2020-01-01 00:45:00
3          302        2020-01-01 00:01:00 2020-01-01 00:54:00 2020-01-01 00:59:00
           303        2020-01-01 00:01:00 2020-01-01 00:01:00 2020-01-01 00:06:00
           300        2020-01-01 00:49:00 2020-01-01 00:49:00 2020-01-01 00:54:00
4          403        2020-01-01 00:02:00 2020-01-01 00:21:00 2020-01-01 00:26:00
           400        2020-01-01 00:05:00 2020-01-01 00:05:00 2020-01-01 00:10:00
           402        2020-01-01 00:24:00 2020-01-01 00:24:00 2020-01-01 00:29:00
           401        2020-01-01 00:42:00 2020-01-01 00:42:00 2020-01-01 00:47:00
5          503        2020-01-01 00:31:00 2020-01-01 00:31:00 2020-01-01 00:36:00
           500        2020-01-01 00:39:00 2020-01-01 00:39:00 2020-01-01 00:44:00

In [9]:
sessions['date_time_max_p5_cummax'] = sessions.groupby('visitor_id')['date_time_max_p5'].cummax()
sessions

date_time_min       date_time_max    date_time_max_p5 date_time_max_p5_cummax
visitor_id session_id                                                                                    
1          100        2020-01-01 00:11:00 2020-01-01 00:11:00 2020-01-01 00:16:00     2020-01-01 00:16:00
           102        2020-01-01 00:48:00 2020-01-01 00:48:00 2020-01-01 00:53:00     2020-01-01 00:53:00
           101        2020-01-01 00:52:00 2020-01-01 00:52:00 2020-01-01 00:57:00     2020-01-01 00:57:00
2          200        2020-01-01 00:16:00 2020-01-01 00:18:00 2020-01-01 00:23:00     2020-01-01 00:23:00
           203        2020-01-01 00:37:00 2020-01-01 00:37:00 2020-01-01 00:42:00     2020-01-01 00:42:00
           201        2020-01-01 00:40:00 2020-01-01 00:40:00 2020-01-01 00:45:00     2020-01-01 00:45:00
3          302        2020-01-01 00:01:00 2020-01-01 00:54:00 2020-01-01 00:59:00     2020-01-01 00:59:00
           303        2020-01-01 00:01:00 2020-01-01 00:01:00 2020-01-01 00:06:00     2020-01-01 00:59:00
           300        2020-01-01 00:49:00 2020-01-01 00:49:00 2020-01-01 00:54:00     2020-01-01 00:59:00
4          403        2020-01-01 00:02:00 2020-01-01 00:21:00 2020-01-01 00:26:00     2020-01-01 00:26:00
           400        2020-01-01 00:05:00 2020-01-01 00:05:00 2020-01-01 00:10:00     2020-01-01 00:26:00
           402        2020-01-01 00:24:00 2020-01-01 00:24:00 2020-01-01 00:29:00     2020-01-01 00:29:00
           401        2020-01-01 00:42:00 2020-01-01 00:42:00 2020-01-01 00:47:00     2020-01-01 00:47:00
5          503        2020-01-01 00:31:00 2020-01-01 00:31:00 2020-01-01 00:36:00     2020-01-01 00:36:00
           500        2020-01-01 00:39:00 2020-01-01 00:39:00 2020-01-01 00:44:00     2020-01-01 00:44:00

In [10]:
sessions['date_time_max_p5_cummax_prev'] = sessions.groupby('visitor_id')['date_time_max_p5_cummax'].transform(pd.Series.shift)
sessions

date_time_min       date_time_max    date_time_max_p5 date_time_max_p5_cummax date_time_max_p5_cummax_prev
visitor_id session_id                                                                                                                 
1          100        2020-01-01 00:11:00 2020-01-01 00:11:00 2020-01-01 00:16:00     2020-01-01 00:16:00                          NaT
           102        2020-01-01 00:48:00 2020-01-01 00:48:00 2020-01-01 00:53:00     2020-01-01 00:53:00          2020-01-01 00:16:00
           101        2020-01-01 00:52:00 2020-01-01 00:52:00 2020-01-01 00:57:00     2020-01-01 00:57:00          2020-01-01 00:53:00
2          200        2020-01-01 00:16:00 2020-01-01 00:18:00 2020-01-01 00:23:00     2020-01-01 00:23:00                          NaT
           203        2020-01-01 00:37:00 2020-01-01 00:37:00 2020-01-01 00:42:00     2020-01-01 00:42:00          2020-01-01 00:23:00
           201        2020-01-01 00:40:00 2020-01-01 00:40:00 2020-01-01 00:45:00     2020-01-01 00:45:00          2020-01-01 00:42:00
3          302        2020-01-01 00:01:00 2020-01-01 00:54:00 2020-01-01 00:59:00     2020-01-01 00:59:00                          NaT
           303        2020-01-01 00:01:00 2020-01-01 00:01:00 2020-01-01 00:06:00     2020-01-01 00:59:00          2020-01-01 00:59:00
           300        2020-01-01 00:49:00 2020-01-01 00:49:00 2020-01-01 00:54:00     2020-01-01 00:59:00          2020-01-01 00:59:00
4          403        2020-01-01 00:02:00 2020-01-01 00:21:00 2020-01-01 00:26:00     2020-01-01 00:26:00                          NaT
           400        2020-01-01 00:05:00 2020-01-01 00:05:00 2020-01-01 00:10:00     2020-01-01 00:26:00          2020-01-01 00:26:00
           402        2020-01-01 00:24:00 2020-01-01 00:24:00 2020-01-01 00:29:00     2020-01-01 00:29:00          2020-01-01 00:26:00
           401        2020-01-01 00:42:00 2020-01-01 00:42:00 2020-01-01 00:47:00     2020-01-01 00:47:00          2020-01-01 00:29:00
5          503        2020-01-01 00:31:00 2020-01-01 00:31:00 2020-01-01 00:36:00     2020-01-01 00:36:00                          NaT
           500        2020-01-01 00:39:00 2020-01-01 00:39:00 2020-01-01 00:44:00     2020-01-01 00:44:00          2020-01-01 00:36:00

In [11]:
sessions['group_with_prev'] = sessions.date_time_min <= sessions.date_time_max_p5_cummax_prev
sessions

date_time_min       date_time_max    date_time_max_p5 date_time_max_p5_cummax date_time_max_p5_cummax_prev  group_with_prev
visitor_id session_id                                                                                                                                  
1          100        2020-01-01 00:11:00 2020-01-01 00:11:00 2020-01-01 00:16:00     2020-01-01 00:16:00                          NaT            False
           102        2020-01-01 00:48:00 2020-01-01 00:48:00 2020-01-01 00:53:00     2020-01-01 00:53:00          2020-01-01 00:16:00            False
           101        2020-01-01 00:52:00 2020-01-01 00:52:00 2020-01-01 00:57:00     2020-01-01 00:57:00          2020-01-01 00:53:00             True
2          200        2020-01-01 00:16:00 2020-01-01 00:18:00 2020-01-01 00:23:00     2020-01-01 00:23:00                          NaT            False
           203        2020-01-01 00:37:00 2020-01-01 00:37:00 2020-01-01 00:42:00     2020-01-01 00:42:00          2020-01-01 00:23:00            False
           201        2020-01-01 00:40:00 2020-01-01 00:40:00 2020-01-01 00:45:00     2020-01-01 00:45:00          2020-01-01 00:42:00             True
3          302        2020-01-01 00:01:00 2020-01-01 00:54:00 2020-01-01 00:59:00     2020-01-01 00:59:00                          NaT            False
           303        2020-01-01 00:01:00 2020-01-01 00:01:00 2020-01-01 00:06:00     2020-01-01 00:59:00          2020-01-01 00:59:00             True
           300        2020-01-01 00:49:00 2020-01-01 00:49:00 2020-01-01 00:54:00     2020-01-01 00:59:00          2020-01-01 00:59:00             True
4          403        2020-01-01 00:02:00 2020-01-01 00:21:00 2020-01-01 00:26:00     2020-01-01 00:26:00                          NaT            False
           400        2020-01-01 00:05:00 2020-01-01 00:05:00 2020-01-01 00:10:00     2020-01-01 00:26:00          2020-01-01 00:26:00             True
           402        2020-01-01 00:24:00 2020-01-01 00:24:00 2020-01-01 00:29:00     2020-01-01 00:29:00          2020-01-01 00:26:00             True
           401        2020-01-01 00:42:00 2020-01-01 00:42:00 2020-01-01 00:47:00     2020-01-01 00:47:00          2020-01-01 00:29:00            False
5          503        2020-01-01 00:31:00 2020-01-01 00:31:00 2020-01-01 00:36:00     2020-01-01 00:36:00                          NaT            False
           500        2020-01-01 00:39:00 2020-01-01 00:39:00 2020-01-01 00:44:00     2020-01-01 00:44:00          2020-01-01 00:36:00            False

In [12]:
sessions['sessions_group_id'] = (~sessions.group_with_prev).cumsum()
sessions

date_time_min       date_time_max    date_time_max_p5 date_time_max_p5_cummax date_time_max_p5_cummax_prev  group_with_prev  sessions_group_id
visitor_id session_id                                                                                                                                                     
1          100        2020-01-01 00:11:00 2020-01-01 00:11:00 2020-01-01 00:16:00     2020-01-01 00:16:00                          NaT            False                  1
           102        2020-01-01 00:48:00 2020-01-01 00:48:00 2020-01-01 00:53:00     2020-01-01 00:53:00          2020-01-01 00:16:00            False                  2
           101        2020-01-01 00:52:00 2020-01-01 00:52:00 2020-01-01 00:57:00     2020-01-01 00:57:00          2020-01-01 00:53:00             True                  2
2          200        2020-01-01 00:16:00 2020-01-01 00:18:00 2020-01-01 00:23:00     2020-01-01 00:23:00                          NaT            False                  3
           203        2020-01-01 00:37:00 2020-01-01 00:37:00 2020-01-01 00:42:00     2020-01-01 00:42:00          2020-01-01 00:23:00            False                  4
           201        2020-01-01 00:40:00 2020-01-01 00:40:00 2020-01-01 00:45:00     2020-01-01 00:45:00          2020-01-01 00:42:00             True                  4
3          302        2020-01-01 00:01:00 2020-01-01 00:54:00 2020-01-01 00:59:00     2020-01-01 00:59:00                          NaT            False                  5
           303        2020-01-01 00:01:00 2020-01-01 00:01:00 2020-01-01 00:06:00     2020-01-01 00:59:00          2020-01-01 00:59:00             True                  5
           300        2020-01-01 00:49:00 2020-01-01 00:49:00 2020-01-01 00:54:00     2020-01-01 00:59:00          2020-01-01 00:59:00             True                  5
4          403        2020-01-01 00:02:00 2020-01-01 00:21:00 2020-01-01 00:26:00     2020-01-01 00:26:00                          NaT            False                  6
           400        2020-01-01 00:05:00 2020-01-01 00:05:00 2020-01-01 00:10:00     2020-01-01 00:26:00          2020-01-01 00:26:00             True                  6
           402        2020-01-01 00:24:00 2020-01-01 00:24:00 2020-01-01 00:29:00     2020-01-01 00:29:00          2020-01-01 00:26:00             True                  6
           401        2020-01-01 00:42:00 2020-01-01 00:42:00 2020-01-01 00:47:00     2020-01-01 00:47:00          2020-01-01 00:29:00            False                  7
5          503        2020-01-01 00:31:00 2020-01-01 00:31:00 2020-01-01 00:36:00     2020-01-01 00:36:00                          NaT            False                  8
           500        2020-01-01 00:39:00 2020-01-01 00:39:00 2020-01-01 00:44:00     2020-01-01 00:44:00          2020-01-01 00:36:00            False                  9

In [14]:
hits = pd.merge(left=hits, right=sessions[['sessions_group_id']], on='session_id')
hits

visitor_id  session_id           date_time                      page_url  sessions_group_id
0            1         100 2020-01-01 00:11:00  shoesfordogs.com/product/apx                  1
1            1         102 2020-01-01 00:48:00  shoesfordogs.com/product/evu                  2
2            1         101 2020-01-01 00:52:00  shoesfordogs.com/product/iev                  2
3            2         200 2020-01-01 00:16:00  shoesfordogs.com/product/aur                  3
4            2         200 2020-01-01 00:18:00  shoesfordogs.com/product/oma                  3
5            2         203 2020-01-01 00:37:00  shoesfordogs.com/product/rgy                  4
6            2         201 2020-01-01 00:40:00  shoesfordogs.com/product/yck                  4
7            3         302 2020-01-01 00:01:00  shoesfordogs.com/product/dah                  5
8            3         302 2020-01-01 00:33:00  shoesfordogs.com/product/pys                  5
9            3         302 2020-01-01 00:48:00  shoesfordogs.com/product/kef                  5
10           3         302 2020-01-01 00:54:00  shoesfordogs.com/product/tfg                  5
11           3         303 2020-01-01 00:01:00  shoesfordogs.com/product/lzj                  5
12           3         300 2020-01-01 00:49:00  shoesfordogs.com/product/kef                  5
13           4         403 2020-01-01 00:02:00  shoesfordogs.com/product/lzj                  6
14           4         403 2020-01-01 00:21:00  shoesfordogs.com/product/oma                  6
15           4         400 2020-01-01 00:05:00  shoesfordogs.com/product/pys                  6
16           4         402 2020-01-01 00:24:00  shoesfordogs.com/product/apx                  6
17           4         401 2020-01-01 00:42:00  shoesfordogs.com/product/evu                  7
18           5         503 2020-01-01 00:31:00  shoesfordogs.com/product/dah                  8
19           5         500 2020-01-01 00:39:00  shoesfordogs.com/product/apx                  9